pkg/bgpv1/annotations.go

```diff
@@ -0,0 +1,185 @@
+// Where {asn} is replaced by the local ASN number of the virtual router.
+//
+// Currently supported attributes are:
+//  router-id=IPv4 (string): when present on a specific node, use this value for
+//                           the router ID of the virtual router with local {asn}
+//  local-port=port (int):  the local port to listen on for incoming BGP connections
+type Attributes struct {
+	// The local ASN of the virtual router these Attributes targets.
+	ASN int
+	// The router ID to use for the virtual router with the above local ASN.
+	RouterID string
+	// The local BGP port to listen on.
+	LocalPort int
+}
+
+// AnnotationMap coorelates a parsed Annotations structure with the local
+// ASN its annotating.
+type AnnotationMap map[int]Attributes
+
+// ErrMulti holds multiple errors and formats them sanely when printed.
+type ErrMulti struct {
+	errs []error
+}
+
+func (e ErrMulti) Error() string {
+	s := strings.Builder{}
+	for _, err := range e.errs {
+		s.WriteString(err.Error() + ",")
+	}
+	return s.String()
+}
+
+// NewAnnotationMap parses a Kubernetes v1.Node object's annotations field into
+// a AnnotationMap and returns the latter.
+//
+// An error is returned containing one or more parsing errors.
+//
+// This is for convenience so the caller can log all parsing errors at once.
+// The error should still be treated as a normal descrete error and an empty
+// AnnotationMap is returned.
+func NewAnnotationMap(n *v1.Node) (AnnotationMap, error) {
+	am := AnnotationMap{}
+	errs := make([]error, 0, len(n.Annotations))
+	for key, value := range n.Annotations {
+		asn, attrs, err := parseAnnotation(key, value)
+		if err != nil && !errors.As(err, &ErrNotVRouterAnno{}) {
+			errs = append(errs, err)
+			continue
+		}
+		am[asn] = attrs
+	}
+	if len(errs) > 0 {
+		return am, ErrMulti{errs}
+	}
+	return am, nil
+}
+
+// parseAnnotation will attempt to parse a `cilium.io/bgp-virtual-router`
+// annotation into an Attributes structure.
+//
+// Errors returned by this parse method are defined at top of file.
+func parseAnnotation(key string, value string) (int, Attributes, error) {
+	var out Attributes
+	// is this an annotation we care about?
+	if !strings.HasPrefix(key, BGPVRouterAnnoPrefix) {
+		return 0, out, ErrNotVRouterAnno{key}
+	}
+
+	// parse out asn from annotation key, if split at "." will be 3rd element
+	var asn int
+	if anno := strings.Split(key, "."); len(anno) != 3 {
+		return 0, out, ErrNoASNAnno{key}
+	} else {
+		var err error
+		asn, err = strconv.Atoi(anno[2])
+		if err != nil {
+			return 0, out, ErrASNAnno{}
+		}
+	}
+	out.ASN = asn
+
+	// split annotation value into multiple "key=value" formatted attributes.
+	attrs := strings.Split(value, ",")
+	if len(attrs) == 0 {
+		return 0, out, nil // empty attributes, not an error
+	}
+	// parse string attributes into Attributes structure.
+	for _, attr := range attrs {
+		kv := strings.Split(attr, "=")
+		if len(kv) != 2 {
+			continue
+		}
+		switch kv[0] {
+		case "router-id":
```

Comment:
In the current implementation, when users specify two `router-id` KV, later one wins. I think it would be nice to document this behavior.